docs: update RFC-0201/TariScript: Add metadata ComSig to the transaction output (#3070)
tari-project/tari: RFC/src/RFC-0201_TariScript.md

$$
\begin{aligned}
C_i = v_i \cdot H  + k_i \cdot G
\end{aligned}
\tag{1}
$$

The [metadata signature] is an aggregated Commitment Signature signed with a combination of the homomorphic 
commitment private values \\( (v\_i \\, , \\, k\_i )\\), with the spending key only known by the receiver, and sender 
offset private key \\(k\_{Oi}\\), only known by the sender. The signature challenge consists of all the transaction 
output metadata, effectively forming a contract between the sender and receiver, making all those values non-malleable 
and ensuring only the sender and receiver can enter into this contract. (See [Signature on Commitment values] by F. 
Zhang et. al. and [Commitment Signature] by G. Yu. for details about this signature.)

Note that the [Commitment Signature] is an aggregated signature between the sender and receiver, which is 
constructed as follows.

The sender portion of the public nonce is:

$$
\begin{aligned}
R_{MSi} &= r_{MSi_a} \cdot H + r_{MSi_b} \cdot G
\end{aligned}
\tag{2}
$$

> The sender sends \\(K\_{Oi}, R_{MSi}\\) to the receiver, whom now have all the information to calculate the final 

Reviewer: ```suggestion
The sender sends \\(K\_{Oi}, R_{MSi}\\) to the receiver, who now has all the required information to calculate the final 
```

$$

The receiver sends \\( s_{MRi} = (a_{MRi}, b_{MRi}, R_{Mi} ) \\) along with the other partial transaction information to 
the sender. The sender starts by calculating the final challenge (16) and then completes their part of the aggregated 
Commitment Signature.

$$
\begin{aligned}
a_{MSi} &= r_{MSi_a} \\\\
b_{MSi} &= r_{MSi_b} + e(k\_{Oi})
\end{aligned}
\tag{6}
$$

Note that (6) is a slight deviation to the Commitment Signature due to the fact that we do not have a private value for 
\\( a_{MSi} \\). This is equivalent to having a commitment to the value of zero (i.e. 
\\( C_i = 0 \cdot H  + k\_{Oi} \cdot G \\)). 

The final aggregated Commitment Signature is combined a follows:

$$
\begin{aligned}
s_{Mi} &= (a_{Mi}, b_{Mi}, R_{Mi} ) \\\\
&= ((a_{MSi} + a_{MRi}), (b_{MSi} + b_{MRi}), (R_{MSi} + R_{MRi}) )
\end{aligned}
\tag{7}
$$

This is verified by the following:

$$
\begin{aligned}
a_{Mi} \cdot H + b_{Mi} \cdot G = R_{Mi} + (C_i + K_{Oi})e
\end{aligned}
\tag{8}
$$

However, when evaluating (8) it is evident that the receiver can calculate \\( r_{MSi_a} \\) as follows:

$$
\begin{aligned}
a_{MSi} &= a_{Mi} - a_{MRi} \\\\
r_{MSi_a} &= a_{MSi}
\end{aligned}
\tag{9}
$$

> To not leak any private nonce from the sender to the receiver, \\( r_{MSi_a} \\) can be set equal to zero, effectively 

Reviewer: ```suggestion
To not leak any private nonces from the sender to the receiver, \\( r_{MSi_a} \\) can be set to zero, effectively 
```

turning the sender portion of the aggregated signature into a normal Schnorr signature. Therefore, (2), (6) and (7) can 
be rewritten as follows:

$$
\begin{aligned}
R_{MSi} &= r_{MSi_b} \cdot G
\end{aligned}
\tag{10}
$$

$$
\begin{aligned}
a_{MSi} &= 0 \\\\
b_{MSi} &= r_{MSi_b} + e(k\_{Oi})
\end{aligned}
\tag{11}
$$
 
$$
\begin{aligned}
s_{Mi} &= (a_{Mi}, b_{Mi}, R_{Mi} ) \\\\
&= (a_{MRi}, (b_{MSi} + b_{MRi}), (R_{MSi} + R_{MRi}) )
\end{aligned}
\tag{12}
$$

Note that:
* The UTXO has a positive value `v` like any normal UTXO.
* The script and the output features can no longer be changed by the miner or any other party. This includes the sender 

>   and receiver; they would need to cooperate to enter into a new contract to change any maetadata, otherwise the 

Reviewer: ```suggestion
  and receiver; they would need to cooperate to enter into a new contract to change any metadata, otherwise the 
```

    script: Vec<u8>,
    /// The script input data, if any
    input_data: Vec<u8>,
    /// Signature signing the script, input data, [script public key] and the homomorphic commitment with a combination 
    /// of the homomorphic commitment private values (amount and blinding factor) and the [script private key].
    script_signature: CommitmentSignature,
    /// The sender offset pubkey, K_O
    sender_offset_public_key: PublicKey
}
```

> The `script_signature` is an aggregated Schnorr signature signed with a combination of the homomorphic commitment 

Reviewer 2: Its not a schnorr signature, its a commitment signature.

> private values \\( (v\_i \\, , \\, k\_i )\\) and [script private key] \\(k\_{Si}\\) to prove ownership of thereof. It 

Reviewer: ```suggestion
private values \\( (v\_i \\, , \\, k\_i )\\) and [script private key] \\(k\_{Si}\\) to prove ownership thereof. It 
```

Verify that for every valid transaction or block:

> 1. The [metadata signature] \\( (a_{Ma}, b_{Ma}, R_{Ma}) \\) is valid for every output.

Reviewer 2: I would rather stick with `s\_{Mi}` as the metadata signature then starting to point to wards its sub parts. Its more consistent with the rest of the document and easier to read in my mind. We should already have explained that `s\_{Mi} = (a_{Ma}, b_{Ma}, R_{Ma})`. 

For example just below this in "4" we talk of just `\\( s\_{Si} \\),` Everywhere else in the document we just talk of the signature as a single variable.

### Standard MW transaction

For this use case we have Alice who sends Bob some Tari.
Bob's wallet is  online and is able to countersign the transaction.

Alice creates a new transaction spending \\( C\_a \\) to a new output containing the commitment \\( C\_b \\) (ignoring fees for now).

To spend \\( C\_a \\), she provides:

* An input that contains \\( C\_a \\).
* The script input, \\( \input_a \\).
* A valid script signature, \\( s\_{Si} \\) as per (13),(14) proving that she owns the commitment 
  \\( C\_a \\), knows the private key, \\( k_{Sa} \\), corresponding to \\( K_{Sa} \\), the public key left on the stack 
  after executing \\( \script_a \\) with \\( \input_a \\).
* A sender offset public key, \\( K_{Ob} \\).
* The script offset, \\( \so\\) with:
$$
\begin{aligned}
\so  = k_{Sa} - k_{Ob}
\end{aligned}
\tag{20}
$$

> Alice sends the usual first round data to Bob. Bob can then completes his side of the transaction and returns his commitment 

Reviewer: ```suggestion
Alice sends the usual first round data to Bob. Bob can then complete his side of the transaction and return his commitment 
```

\\(C\_b\\) along with its rangeproof, the partial signature and public blinding factor back to Alice. All this happens as per the 
[standard Mimblewimble protocol]. 

Because Alice is creating the transaction, she has a final say over the script \\( \script_b \\), similar to a 
[bitcoin transaction], but Bob can also opt to send Alice a script \\(\script\_b\\) of his choosing. If Bob did not send a script, 

> she chooses something akin to a `NOP` script for the script \\(\script\_b\\).

Reviewer: How does Bob know the script that was ultimately used here?

Author: Good pickup; I did not fully complete this section according to the protocol described in (1) to (12).

### Standard MW transaction

For this use case we have Alice who sends Bob some Tari.
Bob's wallet is  online and is able to countersign the transaction.

Alice creates a new transaction spending \\( C\_a \\) to a new output containing the commitment \\( C\_b \\) (ignoring fees for now).

To spend \\( C\_a \\), she provides:

* An input that contains \\( C\_a \\).
* The script input, \\( \input_a \\).

> * A valid script signature, \\( (a_{Sa}, b_{Sa}, R_{Sa}) \\) as per (13),(14) proving that she owns the commitment 

Reviewer 2: Here we talk about the tuple of the signature again... We should be consistent. My vote is in favor of a single value.

  \\( C\_a \\), knows the private key, \\( k_{Sa} \\), corresponding to \\( K_{Sa} \\), the public key left on the stack 
  after executing \\( \script_a \\) with \\( \input_a \\).
* A sender offset public key, \\( K_{Ob} \\).
* The script offset, \\( \so\\) with:
$$
\begin{aligned}
\so  = k_{Sa} - k_{Ob}
\end{aligned}
\tag{20}
$$

Alice sends the usual first round data to Bob. Bob can then completes his side of the transaction and returns his commitment 
\\(C\_b\\) along with its rangeproof, the partial signature and public blinding factor back to Alice. All this happens as per the 
[standard Mimblewimble protocol]. 

Because Alice is creating the transaction, she has a final say over the script \\( \script_b \\), similar to a 
[bitcoin transaction], but Bob can also opt to send Alice a script \\(\script\_b\\) of his choosing. If Bob did not send a script, 
she chooses something akin to a `NOP` script for the script \\(\script\_b\\).

Alice calculates her portion of the [metadata signature] \\( (a_{Ma}, b_{Ma}, R_{Ma}) \\) with: 

> **Hansie to fix**

Reviewer 2: ?

$$
\begin{aligned}
s_{Mb} = r_{mb} + k_{Ob} \hash{ \script_b \cat F_b \cat R_{Mb} }
\end{aligned}
\tag{21}
$$

Alice then adds the following info to Bob's TransactionOutput:

> * Meta_signature \\(s_{Mb}\\),

Reviewer 2: Here it is a single value again.

out-of-band, and might typically be Bob's wallet public key on the Tari network.

Bob requires the value \\( v_b \\) and blinding factor \\( k_b \\) to claim his payment, but he needs to be able to 
claim it without asking Alice for them.

This information can be obtained by using Diffie-Hellman and Bulletproof rewinding. If the blinding factor \\( k\_b \\) 
was calculated with Diffie-Hellman using the sender offset keypair, (\\( k\_{Ob} \\),\\( K\_{Ob} \\)) as the sender 
keypair and the script keypair, \\( (k\_{Sb} \\),\\( K\_{Sb}) \\) as the receiver keypair, the blinding factor 
\\( k\_b \\) can be securely calculated without communication.

Alice uses Bob's public key to create a shared secret, \\( k\_b \\) for the output commitment, \\( C\_b \\), using
Diffie-Hellman key exchange.

Alice calculates \\( k_b \\) as
$$
\begin{aligned}
k_b = k_{Ob} * K_{Sb}
\end{aligned}
\tag{25}
$$

Next Alice next uses Bulletproof rewinding, see [RFC 180](RFC-0180_BulletproofRewinding.md), to encrypt the value 
\\( v_b \\) into the the Bulletproof for the commitment \\( C_b \\). For this she uses 

> \\( k_{rewind} =  \hash{k_{b}} \\) as the rewind_key and \\( k_{blinding} =  \hash{\hash{k_{b}}} \\) as the blinding 

Reviewer: Why can't we use `k_b` as the rewind/blinding key? (I forget)

Author: Considering [RFC-0180](https://rfc.tari.com/RFC-0180_BulletproofRewinding.html).

The security of the Bulletproof range proof in equations (2) and (3) are dependent on the condition that the five blinding factors used there should be sampled randomly from `Z_p`. If we were to only use `k_b`, it will result in using the same rewind nonce for `tilde(a)_j` and `tilde(s)_j` and the same blinding nonce for `tilde(t)_1j` and `tilde(t)_2j`, something that Dalek specifically commented on not to do.

Reviewer: Ah thanks for the explanation :)

| sender offset public key | \\( K_{Oa} \\)                        | Not used in this transaction                                                                  |

| Transaction output | Symbols                               | Knowledge                                                              |
|---------------------------|---------------------------------------|------------------------------------------------------------|
| commitment                | \\( C_b = k_b \cdot G + v \cdot H \\) | Alice and Bob know the blinding factor and value           |
| features                  | \\( F_b \\)                           | Public                                                     |
| script                    | \\( \script_b \\)                     | Script is public. Only Bob knows the correct script input. |
| range proof               |                                       | Alice and Bob know opening parameters                      |
| sender offset public key  | \\( K_{Ob} \\)                        | Alice knows \\( k_{Ob} \\)                                 |

> | metadata signature | \\( s_{Mb}, R_{Mb} \\)                | Alice knows \\( k_{Ob} \\)  and the metadata)              |

Reviewer 2: Just line these up again, they read easier that way.
